File: birthday_problem/birthday_problem.py
from logging import exception
from random import randrange
import statistics

def simulate():
  birthdays = []

  for i in range(1,366):
    birthday = randrange(1, 365, 1)
    if birthday in birthdays:
      return i
    
    birthdays.append(i)

  raise exception('Expectted overlapping birthdays')

def main():
  N = 1_000
  results = list(map(lambda x: simulate(), range(1, N)))
  mean = statistics.mean(results)
  median = statistics.median(results)
  mode =  statistics.mode(results)

  print(f'mean: {mean}, median: {median}, mode: {mode}, N={N}')
main()

File: birthday_problem/birthday_problem.r
get_random_birthday <- function() {
  temp_b <- floor(runif(1, min=1, max=366))
  return (temp_b)
}

getmode <- function(v) {
   uniqv <- unique(v)
   uniqv[which.max(tabulate(match(v, uniqv)))]
}

simulate <- function() {
  birthdays <- numeric()

  for (person_count in 1:365) {
    birthday <- get_random_birthday()
    if (is.element(birthday, birthdays)) {
      return (person_count)
    }
    birthdays <- c(birthday, birthdays)
  }

  stop('Expected overlapping birthdays')
}

main <- function() {
  sim_count <- 10000
  results <- numeric(sim_count)

  for (sim in 1:sim_count) {
    sim_result <- simulate()
    results[sim] <- sim_result
  }

  stats <- c(mean(results), median(results), getmode(results), sim_count)
  names(stats) <- c('mean', 'median', 'mode', 'trials')
  return (stats)
}

main()
